### PR TITLE
TINY-7249: changed loading order so content css gets loaded before loading contents

### PR DESCRIPTION
**Pull request by spocke**
Related Ticket:

Description of Changes:
* This changes the load order of things so that the content css gets loaded before the editor is filled with contents. This is to prevent a FOUC happening while loading the editor. It's extra visible with RTC but I think this happens on the non RTC version of tinymce as well. Not sure how to make a unit test for this though since just waiting for the events to fire is not enough to ensure that a flash doesn't occur.

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/core/main/ts/init/InitContentBody.ts [hunk 1]
```diff
@@ -453,17 +460,23 @@ const initContentBody = (editor: Editor, skipWrite?: boolean) => {
   Placeholder.setup(editor);
 
   Events.firePreInit(editor);
+  preInit(editor);
 
   Rtc.setup(editor).fold(() => {
-    preInit(editor);
-  }, (loadingRtc) => {
+    Events.firePostRender(editor);
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/init/InitContentBody.ts:466]: Changing this event like this is pretty much a breaking change. I don't think we can do this as it's moved a reasonable amount later in the process now and we have no idea how many people were relying on it firing at the old location sorry.

File: modules/tinymce/src/core/test/ts/browser/EditorRemoveTest.ts [hunk 2]
```diff
@@ -38,7 +38,7 @@ describe('browser.tinymce.core.EditorRemoveTest', () => {
     await McEditor.pFromHtml('<textarea></textarea>', {
       ...settings,
       setup: (editor: Editor) => {
-        editor.on('LoadContent', () => {
+        editor.on('PreInit', () => {
```
Comment (lnewson) [inline modules/tinymce/src/core/test/ts/browser/EditorRemoveTest.ts:41]: Why was this changed?

**Review by lnewson — CHANGES_REQUESTED** [review]

File: modules/tinymce/src/core/main/ts/init/InitContentBody.ts (hunk 1 above)
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/init/InitContentBody.ts:466]: For context, a simple example of how this may break an integration is if an integration changed the `directionality` in the `PostRender`. Previously that would have taken effect, but it now no longer will since the directionality is read before `PostRender` fires.

File: modules/tinymce/CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 
 ### Fixed
+- Flash of unstyled content while loading the content css was shown. It will now load the content css first then populate the editor #TINY-7249
```
Comment (lnewson) [inline modules/tinymce/CHANGELOG.md:10]: If we're going ahead with this, then this really needs to be a `Changed` entry as well. We really need to call out/detail that we're changing the initialization sequence here as that could break some people integrations.

**Comment by lnewson** [comment]
Thinking about this more, I think the better/safer/least non-breaking option would be to do something similar to what we do in the theme. Whereby we set `visibility: hidden` while the editor is loading and then remove that once the content CSS has loaded.

File: modules/tinymce/src/core/test/ts/browser/EditorRemoveTest.ts (hunk 2 above)
Comment (spocke) [reply modules/tinymce/src/core/test/ts/browser/EditorRemoveTest.ts:41]: Since it patches over the style sheet loader it needs to be done before the actual css starts loading.

File: modules/tinymce/src/core/main/ts/init/InitContentBody.ts (hunk 1 above)
Comment (spocke) [reply modules/tinymce/src/core/main/ts/init/InitContentBody.ts:466]: Good point let me see if I can move things around so that the rtc instance is setup before preInit and so the we can do all the steps in the exact same order. You can pull the "breaking change card" for any change it's all about how much impact it might have on users. The 5.2, 5.3 change from sync to async init of the content of the editor was far more "breaking" than this re-ordering and it doesn't seem to have broken any integrations. We really need to fix this rendering look at the attached video of the editor blinking it looks really unprofessional. The firePostRender needs to happen early on other wise you get an editor without a toolbar then after a while you get the toolbar and the content so it produces a new type of flash.

Comment (lnewson) [reply modules/tinymce/src/core/main/ts/init/InitContentBody.ts:466]: > You can pull the "breaking change card" for any change it's all about how much impact it might have on users.

Yeah, it has a low risk of breaking someone's integration, however I don't believe we need to do it in this case. We can either change things to make sure the event fires in the same order as you mentioned or look at doing the `hidden` trick I mentioned. If neither of those work though, then yeah we'll have to sort it out and potentially end up changing where things fire.

> The 5.2, 5.3 change from sync to async init of the content of the editor was far more "breaking" than this re-ordering and it doesn't seem to have broken any integrations

I'm going to have to disagree here, as integrations had no hooks into the part of the init sequence that was changed and instead they could only listen for the events we fire. Additionally, the whole init sequence has always been async, since it needs to load stylesheets/javascript files.

Comment (spocke) [reply modules/tinymce/src/core/main/ts/init/InitContentBody.ts]: > I'm going to have to disagree here, as integrations had no hooks into the part of the init sequence that was changed and instead they could only listen for the events we fire. Additionally, the whole init sequence has always been async, since it needs to load stylesheets/javascript files.

Right it was only the RTC part that got converted into an async operation I thought it was both of the branches.

I moved things around now so the postRender event should now fire where it used to fire. Not sure what makes more sense here using a thunk or having a separate function that setups the noop rtcInstance.

File: modules/tinymce/CHANGELOG.md (hunk 3 above)
Comment (spocke) [reply modules/tinymce/CHANGELOG.md]: Added a change item as well since it's a behavior change for sure. Who knows what people might be doing out there.

**Comment by spocke** [comment]
> Thinking about this more, I think the better/safer/least non-breaking option would be to do something similar to what we do in the theme. Whereby we set `visibility: hidden` while the editor is loading and then remove that once the content CSS has loaded.

That would not show a throbber while the content is loading. The RTC side could take a while to initialize so the UI needs to render before content and before content css.

**Comment by lnewson** [comment]
> That would not show a throbber while the content is loading. The RTC side could take a while to initialize so the UI needs to render before content and before content css.

Wait why not? The throbber is UI based so it relies on the skin styles. It should not be impacted by anything we do on the editor body/content css.

File: modules/tinymce/src/core/main/ts/init/InitContentBody.ts (hunk 1 above)
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/init/InitContentBody.ts]: Thanks Spocke! At a glance that should be fine now, but I'll review it properly in the morning.

**Comment by spocke** [comment]
> Wait why not? The throbber is UI based so it relies on the skin styles. It should not be impacted by anything we do on the editor body/content css.

Doesn't the whole thing become a race condition if we stick hidden on the UI then the UI will not show until the content css is loaded but what happens if the UI css gets loaded before content CSS then that will flag the hidden state off by it's !important css rule. We would have to wait the the UI css and the content css to both resolve then we could lift the hidden state by using code then so we would have to tweak the oxide css. This all feels like a refactoring for v6 content css should really be loaded by the silver theme in my opinion also the visibility state should be handled by code not css and !important rules that way it could wait for all css to be loaded in parallel then render the UI. It was handled this way since there was no reliable way to have a load event on CSS on WebKit so the display part of the UI had to be in the css not in code this is no longer the case it been fixed years ago.

File: modules/tinymce/src/core/main/ts/init/InitContentBody.ts (hunk 1 above)
Comment (spocke) [reply modules/tinymce/src/core/main/ts/init/InitContentBody.ts]: This would still need the firePostRender before content css loading change since otherwise we still get that half rendered UI while the RTC content is loading I want the full editor UI in readonly mode with a throbber while the content is loading and that is what is happening now. We could change this to use a hidden state on the body but that would still require the other changes to the loading to avoid that for RTC and it would be a bit more involved that for example in inline mode we don't want to toggle the hidden state on/off since that would flash the editor contents while loading the css in and the page css might already have all the styles it need to render the content. So I think we are good with the things as is for now.

**Review by HAFRMO — APPROVED** [review]

**Review by TheSpyder — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/init/InitContentBody.ts (hunk 1 above)
Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/init/InitContentBody.ts]: I think the thunk is fine.

Whether the UI displays with a toolbar or not is less important than the throbber - that must display immediately. If we can order the events (or safely change other parts of the code) such that the toolbar shows, that's cool, if not that's fine too.

**Review by lnewson — APPROVED** [review]

**Comment by lnewson** [comment]
> Now I think I understand you are talking about setting a visibility: hidden on the body and flag that on/off while rendering. I thought you where talking about this: https://github.com/tinymce/tinymce/blob/develop/modules/oxide/src/less/theme/globals/global.less#L28 having the whole UI switched off then switched on when the content css is loaded that would obviously not render a UI.

Doesn't really matter now, as these changes should work fine, but yes this is what I was talking about.

File: modules/tinymce/src/core/main/ts/Rtc.ts [hunk 4]
```diff
@@ -257,23 +256,24 @@ const getRtcSetup = (editor: Editor): Optional<() => Promise<RtcRuntimeApi>> =>
...
-        editorCast.rtcInstance = makeNoopAdaptor();
-        return Promise.reject<boolean>(err);
-      })
-    )
+    (setup) => {
+      // We need to provide a noop adaptor while initializing since any call by the theme or plugins to say undoManager.hasUndo would throw errors
+      editorCast.rtcInstance = makeNoopAdaptor();
+
+      return Optional.some(
+        () => setup().then((rtcEditor) => {
```
Comment (metricjs) [inline modules/tinymce/src/core/main/ts/Rtc.ts:271]: Just checking: what happened to the `err` here? I assume we no longer need it for some reason, but I couldn't figure out the logic.

**Review by metricjs — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/Rtc.ts (hunk 4 above)
Comment (spocke) [reply modules/tinymce/src/core/main/ts/Rtc.ts:271]: The errors gets handled later on. This specific error branch was added there to provide a noop rtc instance if an error was produced that is now replaced with the noop rtc instance that is always there while initializing but replaced with a real rtc instance once the editor has started.
